src/coreclr/jit/importer.cpp

```diff
@@ -5348,40 +5363,93 @@ GenTree* Compiler::impTypeIsAssignable(GenTree* typeTo, GenTree* typeFrom)
     //
     // to true/false
 
-    if (typeTo->IsCall() && typeFrom->IsCall())
+    // make sure both arguments are `typeof()`
+    CORINFO_CLASS_HANDLE hClassTo   = NO_CLASS_HANDLE;
+    CORINFO_CLASS_HANDLE hClassFrom = NO_CLASS_HANDLE;
+    if (gtIsTypeof(typeTo, &hClassTo) && gtIsTypeof(typeFrom, &hClassFrom))
     {
-        // make sure both arguments are `typeof()`
-        CORINFO_METHOD_HANDLE hTypeof = eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE);
-        if ((typeTo->AsCall()->gtCallMethHnd == hTypeof) && (typeFrom->AsCall()->gtCallMethHnd == hTypeof))
+        TypeCompareState castResult = info.compCompHnd->compareTypesForCast(hClassFrom, hClassTo);
+        if (castResult == TypeCompareState::May)
         {
-            assert((typeTo->AsCall()->gtArgs.CountArgs() == 1) && (typeFrom->AsCall()->gtArgs.CountArgs() == 1));
-            CORINFO_CLASS_HANDLE hClassTo =
-                gtGetHelperArgClassHandle(typeTo->AsCall()->gtArgs.GetArgByIndex(0)->GetEarlyNode());
-            CORINFO_CLASS_HANDLE hClassFrom =
-                gtGetHelperArgClassHandle(typeFrom->AsCall()->gtArgs.GetArgByIndex(0)->GetEarlyNode());
+            // requires runtime check
+            // e.g. __Canon, COMObjects, Nullable
+            return nullptr;
+        }
 
-            if (hClassTo == NO_CLASS_HANDLE || hClassFrom == NO_CLASS_HANDLE)
-            {
-                return nullptr;
-            }
+        GenTreeIntCon* retNode = gtNewIconNode((castResult == TypeCompareState::Must) ? 1 : 0);
+        impPopStack(); // drop both CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE calls
+        impPopStack();
 
-            TypeCompareState castResult = info.compCompHnd->compareTypesForCast(hClassFrom, hClassTo);
-            if (castResult == TypeCompareState::May)
-            {
-                // requires runtime check
-                // e.g. __Canon, COMObjects, Nullable
-                return nullptr;
-            }
+        return retNode;
+    }
 
-            GenTreeIntCon* retNode = gtNewIconNode((castResult == TypeCompareState::Must) ? 1 : 0);
-            impPopStack(); // drop both CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE calls
-            impPopStack();
+    return nullptr;
+}
 
-            return retNode;
+GenTree* Compiler::impTypeGetTypeCode(GenTree* type)
+{
+    GenTree*             retNode = nullptr;
+    CORINFO_CLASS_HANDLE hClass  = NO_CLASS_HANDLE;
+    if (gtIsTypeof(type, &hClass))
+    {
+        int typeCode = -1;
+        switch (info.compCompHnd->getTypeForPrimitiveValueClass(hClass))
```

Comment:
I am not sure how I feel about having TypeCode intrinsic in the JIT. We are saying that TypeCode is a legacy and at the same time adding it to the JIT as the core intrinsic that will encourage more use.

Would it make sense to flip this and build `Enum.GetUnderlyingType` as the core JIT intrinsic, and make sure that it works great with "battery of if" checkes that we use everywhere else?

And if we want to make GetTypeCode better, it should be still possible to do it in the C# using IsKnownConstant. It would avoid hardcoding these constants here and it would allow us to handle all cases, including the ones that are not handled here currently.

Reply:
TypeCode is legacy, but there is a single place it's still useful - enums, which the exisiting implementation uses in a few places and changing that could have negative performance impact on runtimes like Mono.

Truth is that I originally tried to make `Enum.GetUnderlyingType` an intrinsic, but I've repurposed the code after being told that there is no way to create type handles I'd need to return in the importer without adding JIT-EE APIs, so I've decided that making `GetUnderlyingType` use the GetTypeCode intrinsic in C# will be much simpler.

And since like I said, the only reasonable usecase for GetTypeCode is enums, it'd not be possible to handle it reasonably with IsKnownConstant without intrinsifying GetUnderlyingType which like I said is tricky. And at the same time, other cases there don't really matter since they can't happen with enums (other than nint/nuint, but those are currently impossible to define in C#).

Reply:
this is currently +489 −66 lines of change with literally zero asm-diffs https://dev.azure.com/dnceng/public/_build/results?buildId=1881778&view=ms.vss-build-web.run-extensions-tab. was this the expected outcome of adding three jit intrinsics? there should be at least some codegen improvement to justify this kind of a change. none of three intrinsics added here are helping the common code patterns.

i can imagine there are many other apis and code patterns which can use intrinsic help, with your time, knowledge and effort.

Reply:
Those intrinsics do affect things, for example Enum.TryParse uses [both](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Enum.cs#L679) of the methods intrinsified here, I have no idea why this would show zero diffs as I have verified that the intrinsics are working locally.

Reply:
> enums, which the exisiting implementation uses in a few places and changing that could have negative performance impact on runtimes like Mono.

The current Enums implementation is a mix CorElementType and TypeCode for historic reasons. I would rather standardize on CorElementType that is more natural runtime representation than on TypeCode. Standardizing the Enum implementation on CorElementType should not have negative impact on Mono.

> for example Enum.TryParse uses both of the methods intrinsified here, I have no idea why this would show zero diffs

It means that we are not using this method anywhere in the framework. It is not that surprising since the generic enum methods have questionable performance characteristics - a lot of code bloat due to generic expansion, but still pretty slow for how big code it generates.

Reply:
The issue is that CorElementType from what I've seen is only obtainable from instance methods and it's internal, external code has no access to it which I'd say still warranties optimizing this.

As for the diffs, @EgorBo told me that SPMI likely won't see changes from such intrinsics, only general JIT changes, and that I'd need to run jit-diffs, which I'll do after #72136 gets merged. I'd personally expect things like enum parsing to be heavily used by 3rd party code.

Reply:
AFAIK, the only time we need to run jit-diff manually (and do the [jit-analyze step](https://github.com/am11/CrossRepoCITesting/blob/master/.github/workflows/jit-diff.yml#L150)) is when we have changes in JIT interface (and `JITEEVersionIdentifier` is updated). Otherwise the superpmi-diffs leg captures the diffs reliably in CI. It is more likely what @jkotas has mentioned that this means the code path in question is unused/dead.

Reply:
> I'd personally expect things like enum parsing to be heavily used by 3rd party code.

It would be useful to find examples.

Reply:
The best example would be all kinds of text based data storage and interchange formats like JSON and XML, although System.Text.Json uses the non generic overload of TryParse.